ASoC: SOF: ipc4-pcm: fix delay calculation when DSP resamples

kv2019i opened the pull request:
When the sampling rates going in (host) and out (dai) from the DSP are different, the IPC4 delay reporting does not work correctly. Add support for this case by scaling the host position values to a common timebase before calculating real-time delay for the PCM.

Link: https://github.com/thesofproject/linux/issues/5498

kv2019i committed:
ASoC: SOF: ipc4-pcm: fix delay calculation when DSP resamples

When the sampling rates going in (host) and out (dai) from the DSP
are different, the IPC4 delay reporting does not work correctly.
Add support for this case by scaling the all raw position values to
a common timebase before calculating real-time delay for the PCM.

Fixes: 0ea06680dfcb ("ASoC: SOF: ipc4-pcm: Correct the delay calculation")
Signed-off-by: Kai Vehmanen <kai.vehmanen@linux.intel.com>

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -1100,19 +1101,71 @@ static int sof_ipc4_get_stream_start_offset(struct snd_sof_dev *sdev,
+	return value;
+}
+
+static u64 sof_ipc4_time_dai_to_host(struct sof_ipc4_timestamp_info *time_info, u64 dai_time)
```
`dai_time` is is not time, but number of frames, name it as 'frames' ?

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -1100,19 +1101,71 @@ static int sof_ipc4_get_stream_start_offset(struct snd_sof_dev *sdev,
+	return sof_ipc4_time_scale(time_info, dai_time, true);
+}
+
+static u64 sof_ipc4_time_host_to_dai(struct sof_ipc4_timestamp_info *time_info, u64 host_time)
```
same here, `host_time` is not time, but `frames`

kv2019i replied:
I coined this as a "sof_ipc4_time_" prefix as this is a helper for timekeeping functionality, but I see you interprted this differently. Will change.

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -1149,8 +1202,13 @@ static int sof_ipc4_pcm_pointer(struct snd_soc_component *component,
 	host_ptr = host_cnt;
 
+	/* Scale value to DAI time in case DAI running at different rate */
+	host_cnt = sof_ipc4_time_host_to_dai(time_info, host_cnt);
```
Why we don't scale the dai_cnt to host time-scale instead and save on back-forth scaling (here host to dai and later from dai to host)?

kv2019i replied:
Hmm, true, now that we can't assume which will wrap around faster, we might as well use host scale as the defaul and then we can avoid converting the resulting delay. Let me work on this. I have an additional fix in the works as well, will add it to this series.

ujfalusi reviewed (commented):
@kv2019i, this is indeed was not handled correctly, thanks for fixing it!
Few comments, but the patch looks good otherwise. I would add fixes tag if applicable

kv2019i reviewed (commented):
Thanks @ujfalusi , will work on an update.

kv2019i commented:
V3 pushed:
- addressed comments from @ujfalusi 
- one fix/change added to the series, this will filter out bogus delay values when delay is querid during an xrun

kv2019i commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -1265,6 +1265,12 @@ static int sof_ipc4_pcm_pointer(struct snd_soc_component *component,
 
+	if (time_info->delay > (DELAY_BOUNDARY >> 1)) {
+		dev_dbg_ratelimited(sdev->dev, "inaccurate delay, host %llu dai_cnt %llu",
+				    host_cnt, dai_cnt);
```
This gets hit quite often with alsa_conformance_test(.py) and high sampling rates as alsa_conformance_test by default uses very small buffers and default is set in frames, so the amount of buffering between CPU and HW in terms of wall-clock time is short at 48000, but it gets super short at higher rates. This doesn't really interfere with testing, but xruns are sometimes hit and if we report delays of 2^31 frames, this completely confuses alsa_conformance_test. Returning zero in these cases seems more right thing to do. E.g. wth appls like video playback, a crazy high delay value can freeze video playback, while a transient zero might just through the audio track off for a few seconds.

ujfalusi replied:
is this happens on stream start or during a running stream?

With any PCM or only with deepbuffer?

kv2019i replied:
@ujfalusi  Does happen both at start and during runtime. Can be hit with HDMI and nocodec SSP PCMs as well.

ujfalusi replied:
Hrm, does it happen on ChainDMA w/o the first patch?
ChainDMA runs with the same rate and both host and dai position is queried from the host side as FW does not provide this information.
ChainDMA and SSP use cases are different when it comes to delay reporting...

kv2019i replied:
@ujfalusi Yes, it still happens. I get this with alsa_conformance_test and on some machines (not all, not related to SOF in anyway), it just hits xruns and e.g. on playback hw_ptr catches appl_ptr. This is its own problem of course, but when this does happen, we report a crazily large value as delay. As a result, even if there is a single xrun, all the alsa_conformance_rate metrics will be out-of-bounds and test will fail.

If it would be just the alsa_conformance_test, maybe we could ignore and just mark this as an app bug, but I think other apps could hit this as well. E.g. if you ar eplaying back video and there is some odd system glitch that causes a transient xrun, we shouldn't make matters worse by reporting a 2^25 delay and cause the video lip-sync to freeze the screen.

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -1167,6 +1210,8 @@ static int sof_ipc4_pcm_pointer(struct snd_soc_component *component,
 	}
+
+	dai_cnt = sof_ipc4_frames_dai_to_host(time_info, dai_cnt);
 	dai_cnt += time_info->stream_end_offset;
```
I think the offset should be applied before moving the dai counter to host domain.
It is counted on the dai domain.

kv2019i replied:
@ujfalusi I convert the offsets to host rate on L1100 and document the fields in L20.

ujfalusi replied:
OK, I missed that.

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -1099,20 +1130,30 @@ static int sof_ipc4_get_stream_start_offset(struct snd_sof_dev *sdev,
 
+/*
+ * Modulus to use to compare host and link counters. This is required
+ * as host/link counters use different units (bytes/frames) and the
```
The bytes on the host side as unit is not relevant here imho, it is converted to frames, what matters is the fact that the rate might be different and the thus the counter on dai/host might wrap at different time for the same duration.

IOW, you introduce an artificial low and hogh enough wrapping point for both to be used, right?

Previously we moved the wrapping point of the DAI earlier to match with how the host will wrap.

kv2019i replied:
@ujfalusi I left the note on bytes as this modulus calculation is still needed even if there is no difference in rates. 
I can drop it if it confuses too much. UPDATE: I will drop it. The conversion to frames is a separate step, you are right.

The idea is to use a made-up wrap point that is smaller than any possible hw counter (e.g. our link and host DMAs), and is much larger than any possible valid delay (we can express 93min delay at 768khz with the U32_MAX as wrap point).

The calculation is the same whether DAI or host side wraps first.

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -1200,8 +1245,12 @@ static int sof_ipc4_pcm_pointer(struct snd_soc_component *component,
 
-	/* Wrap the dai counter at the boundary where the host counter wraps */
-	div64_u64_rem(dai_cnt, time_info->boundary, &dai_cnt);
+	/* dai/host_cnt converted to same unit, but the values will
```
The dai counter is moved to host rate domain, make sure that they are wrapped at the same value?

kv2019i replied:
@ujfalusi This is what we do here, right? DELAY_BOUNDARY is a adhoc boundary that will work with any sampling rate and we scale both values to same base before comparison.

ujfalusi replied:
I think the 'unit' confuses me, can we use rate domain to leave no room for misunderstanding?

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -1099,20 +1130,30 @@ static int sof_ipc4_get_stream_start_offset(struct snd_sof_dev *sdev,
+ * be smaller than the wrap-around point of any hardware counter, as
+ * expressed in units of the host frame counter.
+ */
+#define DELAY_BOUNDARY		U32_MAX
```
It is preferred to have the define at the top of the file and not embedded in middle

kv2019i replied:
Ack, will move up.

kv2019i reviewed (commented):
Replying to review comments.

kv2019i commented:
V4 uploaded:
- moved DELAY_BOUNDARY def at start
- reworded some of the commentary, please @ujfalusi check if this helps readability

kv2019i commented:
V5:
- dropped the second patch -- let's focus on the SRC case first

kv2019i commented:
V6:
- forgot to fix the stream_start_offset calculation for ChainDMA case. the value is already in host sampling rate (as there can be no SRC with chaindma), so the conversion from from dai_to_host format is bogus (the function would not have done anything given the rates are the same, but this is still wrong)
